JIT Type parser for mobile

iseeyuan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30391 JIT Type parser for mobile**

A Type parser to parse the python string of a Type. For example, 
"Tuple[str, Optional[float], Dict[str, List[Tensor]], int]". 
Please refer to test_type_parser.cpp for the usage. 

One of the use cases is in lite interpreter, types needs to be serialized (directly calling the python_str() of the Type) and deserialized (calling parseType(str)).

Differential Revision: [D18924268](https://our.internmc.facebook.com/intern/diff/D18924268)

Commit:
Type parser

Commit:
Update on "Type parser"

Commit:
Update on "Type parser"

Commit:
Update on "JIT Type parser"


A Type parser to parse the python string of a Type. For example, 
"Tuple[str, Optional[float], Dict[str, List[Tensor]], int]". 
Please refer to test_type_parser.cpp for the usage. 

One of the use cases is in lite interpreter, types needs to be serialized (directly calling the python_str() of the Type) and deserialized (calling parseType(str)).

bhosmer commented:
Don't know if you guys went over this before I started sitting in, but is there any way to reuse (or factor out a common subset of) [SchemaTypeParser](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/script/schema_type_parser.cpp) for this? At a high level it's a bummer to have two, but maybe there are e.g. dependency or compatibility issues I'm not aware of? cc @zdevito

iseeyuan commented:
@bhosmer I was not aware of SchemaTypeParser. Thanks for bringing that up! I'll take a look if we could reuse it.

bhosmer commented:
@iseeyuan sure! Having recently become familiar with it myself, I'm eager to spread the word 😁 LMK what you find - e.g. if it looks like there's opportunity for reuse but it would take some factoring work, we should try to make that happen if we can.

zdevito commented on aten/src/ATen/core/type.cpp:
```diff
@@ -655,4 +655,90 @@ void checkNoAny(const Type& base, const char* what, const std::string& attrname,
       "' but it contains an Any type. Any types cannot be members of modules, classes, or named tuples.");
 }
 
+namespace {
```
Since this is mobile specific we should put it in the mobile folder and name it something to indicate that it is the mobile version of type parser.

zdevito commented on aten/src/ATen/core/type.cpp:
```diff
@@ -655,4 +655,90 @@ void checkNoAny(const Type& base, const char* what, const std::string& attrname,
+ public:
+  explicit TypeParser(const std::string& pythonStr)
+    : pythonStr_(pythonStr), start_(0) {}
+    
```
To be consistent with other parsers in PyTorch, this class should have a `next()` method which returns a `std::string` of the next token, and advances to the next state.

iseeyuan replied:
next() and cur() are added.

zdevito commented on aten/src/ATen/core/type.cpp:
```diff
@@ -655,4 +655,90 @@ void checkNoAny(const Type& base, const char* what, const std::string& attrname,
+    for (; start_ < pythonStr_.size() && !isSpecialChar(pythonStr_[start_]);
+           ++start_) {
+      if (pythonStr_[start_] != ' ') {
+        token += pythonStr_[start_];
```
Char-by-char string creation is going to be n^2. This should only extract the string when it has found the end.

iseeyuan replied:
The logic is changed to index points and integrated in lex() function.

zdevito commented on aten/src/ATen/core/type.cpp:
```diff
@@ -655,4 +655,90 @@ void checkNoAny(const Type& base, const char* what, const std::string& attrname,
+        token += pythonStr_[start_];
+      }
+    }
+    if (token == "int") {
```
@bhosmer - It would be nice to share the same parser logic but there are a couple issues. First, we want to serialize using Python type syntax, not our bespoke schema syntax. So we are actually not using the SchemaTypeParser syntax here, but the syntax used in our normal script parser. That parser parses types into an AST first, which is an extra bit of binary size we might not want in mobile.

bhosmer replied:
@zdevito aha! yeah, I think I haven't fully internalized that there are two type syntaxes :) gotcha re [ScriptTypeParser](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/script/script_type_parser.cpp) (as opposed to SchemaTypeParser) pulling in the AST machinery. (How much binary size do you think it would add?)

I guess I'm generically hoping there's some way to single-source some of the meaningful specifics, even if it's infeasible to just use the parser whole. Short of an infeasibly deep refactoring, maybe there are helpers we could use from both parsers, like https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/script/script_type_parser.cpp#L10? cc @iseeyuan

iseeyuan replied:
@bhosmer Yes we could use the helpers. Thanks for bringing it up! But for the script_type_parser as a whole, it looks like it has dependencies on ir, tree, etc., where lite interpreter does not depend on. If we include those dependencies I'm pretty sure the size will be significantly bigger (around 1M) with no decoupling work.

iseeyuan replied:
The basic string to type lookup table is shared now. For compose types (List, Tuple, Dict, etc.), a template function is used for single-element type. Couldn't find a quick and simple way to make them fully generic and reusable. I'm wondering if we could first make this version work so that we could unblock the operator work in general (which are blocking the automation flow in mobile), and revisit it when the operator work is done.

bhosmer replied:
@iseeyuan I think that's fine. If you wouldn't mind, could you make an issue for common lookup of generic types across these two parsers and assign it to me? I can follow up in parallel.

@zdevito qq on [this comment](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/jit_type.h#L641) - as far as I can tell this comment doesn't reflect current usage, right? Asking bc threading calls through a memo cache of instantiations would be one of the ways to factor out the common code here.

iseeyuan replied:
Issue 31085 has been created and assigned to @bhosmer .

zdevito commented on aten/src/ATen/core/type.cpp:
```diff
@@ -655,4 +655,90 @@ void checkNoAny(const Type& base, const char* what, const std::string& attrname,
+    } else if (token == "Any") {
+      return AnyType::create();
+    } else if (token == "Tensor") {
+      return TensorType::create(
```
This should use `TensorType::get()` instead, which sets the more obscure flags that come at the end of this constructor correctly for default tensors.

iseeyuan replied:
Merged to the look up table, where get() is used.

zdevito commented on aten/src/ATen/core/type.cpp:
```diff
@@ -655,4 +655,90 @@ void checkNoAny(const Type& base, const char* what, const std::string& attrname,
+    } else if (token == "Tuple") {
+      std::vector<TypePtr> types;
+      while (start_ < pythonStr_.size() && pythonStr_[start_] != ']') {
+        ++start_; // skip '[' or ','
```
This should be checking that the tokeen actually is a comma. This does not handle whitespace correctly.

zdevito commented on aten/src/ATen/core/type.cpp:
```diff
@@ -655,4 +655,90 @@ void checkNoAny(const Type& base, const char* what, const std::string& attrname,
+      checkAndInc(']');
+      return TupleType::create(types);
+    } else {
+      AT_ERROR("Type ", token, " is not supported in the parser yet.");
```
This error messages makes it seem like `token` is valid but not supported yet, but it also possibly just wrong. The error message should reflect that.

zdevito commented on test/cpp/jit/test_type_parser.cpp:
```diff
@@ -0,0 +1,22 @@
+  std::string int_tps = int_tp->python_str();
+  ASSERT_EQ(int_ps, int_tps);
+
+  std::string tuple_ps("Tuple[str, Optional[float], Dict[str, List[Tensor]], int]");
```
This should have more tests for error cases: empty string, unmatched brackets, weird whitespace. If you can crash this parser then it is possible to construct a model that can be loaded to a mobile device that would crash the application, which is a security risk.

iseeyuan replied:
Tests for error cases are added.

zdevito reviewed (commented):
Thanks for putting this together. Comments are inline.

Commit:
Update on "JIT Type parser for mobile"


A Type parser to parse the python string of a Type. For example, 
"Tuple[str, Optional[float], Dict[str, List[Tensor]], int]". 
Please refer to test_type_parser.cpp for the usage. 

One of the use cases is in lite interpreter, types needs to be serialized (directly calling the python_str() of the Type) and deserialized (calling parseType(str)).

Commit:
Update on "JIT Type parser for mobile"


A Type parser to parse the python string of a Type. For example, 
"Tuple[str, Optional[float], Dict[str, List[Tensor]], int]". 
Please refer to test_type_parser.cpp for the usage. 

One of the use cases is in lite interpreter, types needs to be serialized (directly calling the python_str() of the Type) and deserialized (calling parseType(str)).

Commit:
Update on "JIT Type parser for mobile"


A Type parser to parse the python string of a Type. For example, 
"Tuple[str, Optional[float], Dict[str, List[Tensor]], int]". 
Please refer to test_type_parser.cpp for the usage. 

One of the use cases is in lite interpreter, types needs to be serialized (directly calling the python_str() of the Type) and deserialized (calling parseType(str)).

Commit:
Update on "JIT Type parser for mobile"


A Type parser to parse the python string of a Type. For example, 
"Tuple[str, Optional[float], Dict[str, List[Tensor]], int]". 
Please refer to test_type_parser.cpp for the usage. 

One of the use cases is in lite interpreter, types needs to be serialized (directly calling the python_str() of the Type) and deserialized (calling parseType(str)).

bhosmer reviewed (commented):
This looks good to me - thanks for factoring out the common ground type lookup! I'll follow up with some factoring of the generics. 

One lexing thing noted inline and a couple more tests, then I think we're good, mod anything @zdevito spots that I've missed :)

bhosmer reviewed (commented):
This looks good to me - thanks for factoring out the common ground type lookup! I'll follow up with some factoring of the generics. 

One lexing thing noted inline and a couple more tests, then I think we're good, mod anything @zdevito spots that I've missed :)

bhosmer commented:
BTW do we know what the test failures are?

zdevito commented on torch/csrc/jit/mobile/type_parser.cpp:
```diff
@@ -0,0 +1,121 @@
+
+  std::string pythonStr_;
+  size_t start_;
+  std::queue<std::string> next_tokens_;
```
This queue never has more than a single element in it, so it seems weird to have it be a queue instead of e.g. an optional<std::string> or even just a string that defaults to "" when there are no more tokens.

zdevito reviewed (approved):
Looks good with one small nit.

Commit:
Update on "JIT Type parser for mobile"


A Type parser to parse the python string of a Type. For example, 
"Tuple[str, Optional[float], Dict[str, List[Tensor]], int]". 
Please refer to test_type_parser.cpp for the usage. 

One of the use cases is in lite interpreter, types needs to be serialized (directly calling the python_str() of the Type) and deserialized (calling parseType(str)).

Commit:
Update on "JIT Type parser for mobile"


A Type parser to parse the python string of a Type. For example, 
"Tuple[str, Optional[float], Dict[str, List[Tensor]], int]". 
Please refer to test_type_parser.cpp for the usage. 

One of the use cases is in lite interpreter, types needs to be serialized (directly calling the python_str() of the Type) and deserialized (calling parseType(str)).

Commit:
Update on "JIT Type parser for mobile"


A Type parser to parse the python string of a Type. For example, 
"Tuple[str, Optional[float], Dict[str, List[Tensor]], int]". 
Please refer to test_type_parser.cpp for the usage. 

One of the use cases is in lite interpreter, types needs to be serialized (directly calling the python_str() of the Type) and deserialized (calling parseType(str)).

iseeyuan commented:
> BTW do we know what the test failures are?

Likely from gtest. I posted it here to ask for solutions: https://fb.workplace.com/groups/1144215345733672/permalink/1497569477064922/.

iseeyuan commented on test/cpp/jit/test_mobile_type_parser.cpp:
```diff
@@ -0,0 +1,44 @@
+
+  std::string mismatch3("Dict[Tensor]");
+  EXPECT_ANY_THROW(c10::parseType(mismatch2));
+}
```
"int int' did capture a bug in existing code. I've fixed it. Thanks!

Commit:
Update on "JIT Type parser for mobile"


A Type parser to parse the python string of a Type. For example, 
"Tuple[str, Optional[float], Dict[str, List[Tensor]], int]". 
Please refer to test_type_parser.cpp for the usage. 

One of the use cases is in lite interpreter, types needs to be serialized (directly calling the python_str() of the Type) and deserialized (calling parseType(str)).

Commit:
Update on "JIT Type parser for mobile"


A Type parser to parse the python string of a Type. For example, 
"Tuple[str, Optional[float], Dict[str, List[Tensor]], int]". 
Please refer to test_type_parser.cpp for the usage. 

One of the use cases is in lite interpreter, types needs to be serialized (directly calling the python_str() of the Type) and deserialized (calling parseType(str)).

Differential Revision: [D18924268](https://our.internmc.facebook.com/intern/diff/D18924268)

Commit:
Update on "JIT Type parser for mobile"


A Type parser to parse the python string of a Type. For example, 
"Tuple[str, Optional[float], Dict[str, List[Tensor]], int]". 
Please refer to test_type_parser.cpp for the usage. 

One of the use cases is in lite interpreter, types needs to be serialized (directly calling the python_str() of the Type) and deserialized (calling parseType(str)).

Differential Revision: [D18924268](https://our.internmc.facebook.com/intern/diff/D18924268)

Commit:
Update on "JIT Type parser for mobile"


A Type parser to parse the python string of a Type. For example, 
"Tuple[str, Optional[float], Dict[str, List[Tensor]], int]". 
Please refer to test_type_parser.cpp for the usage. 

One of the use cases is in lite interpreter, types needs to be serialized (directly calling the python_str() of the Type) and deserialized (calling parseType(str)).

Differential Revision: [D18924268](https://our.internmc.facebook.com/intern/diff/D18924268)

iseeyuan committed:
Update on "JIT Type parser for mobile"


A Type parser to parse the python string of a Type. For example, 
"Tuple[str, Optional[float], Dict[str, List[Tensor]], int]". 
Please refer to test_type_parser.cpp for the usage. 

One of the use cases is in lite interpreter, types needs to be serialized (directly calling the python_str() of the Type) and deserialized (calling parseType(str)).

Differential Revision: [D18924268](https://our.internmc.facebook.com/intern/diff/D18924268)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit ceaecd3:
* **1/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=ceaecd309d417fee7ecb41dd0044c545eaf7c2b7).

### 1 new failure recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3912919) pytorch_linux_xenial_py3_6_gcc5_4_test (1/1)
**Step:** "Test" ([details](https://dr.pytorch.org/build-details.html?build_id=72695559))
```
Dec 12 01:12:59   File "test_nn.py", line 56, in <module>
Dec 12 01:12:59     import hypothesis_utils as hu
Dec 12 01:12:59   File "/var/lib/jenkins/workspace/test/hypothesis_utils.py", line 9, in <module>
Dec 12 01:12:59     from hypothesis.searchstrategy import SearchStrategy
Dec 12 01:12:59 ModuleNotFoundError: No module named 'hypothesis.searchstrategy'
Dec 12 01:12:59 Traceback (most recent call last):
Dec 12 01:12:59   File "test/run_test.py", line 456, in <module>
Dec 12 01:12:59     main()
Dec 12 01:12:59   File "test/run_test.py", line 449, in main
Dec 12 01:12:59     raise RuntimeError(message)
Dec 12 01:12:59 RuntimeError: test_nn failed!
Dec 12 01:12:59 + cleanup
Dec 12 01:12:59 + retcode=1
Dec 12 01:12:59 + set +x
Dec 12 01:12:59 =================== sccache compilation log ===================
Dec 12 01:12:59 =========== If your build fails, please take a look at the log above for possible reasons ===========
Dec 12 01:12:59 Compile requests                  7
Dec 12 01:12:59 Compile requests executed         6
Dec 12 01:12:59 Cache hits                        0
Dec 12 01:12:59 Cache misses                      6
Dec 12 01:12:59 Cache timeouts                    0
```


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

iseeyuan committed:
Update on "JIT Type parser for mobile"


A Type parser to parse the python string of a Type. For example, 
"Tuple[str, Optional[float], Dict[str, List[Tensor]], int]". 
Please refer to test_type_parser.cpp for the usage. 

One of the use cases is in lite interpreter, types needs to be serialized (directly calling the python_str() of the Type) and deserialized (calling parseType(str)).

Differential Revision: [D18924268](https://our.internmc.facebook.com/intern/diff/D18924268)

[ghstack-poisoned]

iseeyuan committed:
Update on "JIT Type parser for mobile"


A Type parser to parse the python string of a Type. For example, 
"Tuple[str, Optional[float], Dict[str, List[Tensor]], int]". 
Please refer to test_type_parser.cpp for the usage. 

One of the use cases is in lite interpreter, types needs to be serialized (directly calling the python_str() of the Type) and deserialized (calling parseType(str)).

Differential Revision: [D18924268](https://our.internmc.facebook.com/intern/diff/D18924268)

[ghstack-poisoned]

iseeyuan committed:
Update on "JIT Type parser for mobile"


A Type parser to parse the python string of a Type. For example, 
"Tuple[str, Optional[float], Dict[str, List[Tensor]], int]". 
Please refer to test_type_parser.cpp for the usage. 

One of the use cases is in lite interpreter, types needs to be serialized (directly calling the python_str() of the Type) and deserialized (calling parseType(str)).

Differential Revision: [D18924268](https://our.internmc.facebook.com/intern/diff/D18924268)

[ghstack-poisoned]

iseeyuan committed:
Update on "JIT Type parser for mobile"


A Type parser to parse the python string of a Type. For example, 
"Tuple[str, Optional[float], Dict[str, List[Tensor]], int]". 
Please refer to test_type_parser.cpp for the usage. 

One of the use cases is in lite interpreter, types needs to be serialized (directly calling the python_str() of the Type) and deserialized (calling parseType(str)).

Differential Revision: [D18924268](https://our.internmc.facebook.com/intern/diff/D18924268)

[ghstack-poisoned]

iseeyuan committed:
Update on "JIT Type parser for mobile"


A Type parser to parse the python string of a Type. For example, 
"Tuple[str, Optional[float], Dict[str, List[Tensor]], int]". 
Please refer to test_type_parser.cpp for the usage. 

One of the use cases is in lite interpreter, types needs to be serialized (directly calling the python_str() of the Type) and deserialized (calling parseType(str)).

Differential Revision: [D18924268](https://our.internmc.facebook.com/intern/diff/D18924268)

[ghstack-poisoned]
